libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt

```diff
@@ -0,0 +1,114 @@
+package com.mapbox.navigation.core.reroute
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.base.common.logger.Logger
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
+import com.mapbox.navigation.core.directions.session.DirectionsSession
+import com.mapbox.navigation.core.directions.session.RoutesRequestCallback
+import com.mapbox.navigation.core.routeoptions.RouteOptionsProvider
+import com.mapbox.navigation.core.trip.session.TripSession
+
+/**
+ * Default implementation of [RerouteController]
+ */
+internal class MapboxRerouteController(
+    private val directionsSession: DirectionsSession,
+    private val tripSession: TripSession,
+    private val routeOptionsProvider: RouteOptionsProvider,
+    private val logger: Logger
+) : RerouteController {
+
+    private val observers = mutableSetOf<RerouteController.RerouteStateObserver>()
+
+    override var state: RerouteState = RerouteState.Idle
+        set(value) {
+            if (field == value) {
+                return
+            }
+            field = value
+            observers.forEach { it.onNewState(field) }
+        }
+
+    private companion object {
+        const val TAG = "MapboxRerouteController"
+    }
+
+    // current implementation ignore `onNewRoutes` callback because `DirectionsSession` update routes internally
+    override fun reroute(routesCallback: RerouteController.RoutesCallback) {
```

Comment:
Where is routesCallback actually used? only place it's referenced in tests, is testing that it was not called. 

```
      // routeCallback mustn't called in current implementation. DirectionSession update routes internally
        verify(exactly = 0) { routeCallback.onNewRoutes(any()) }
    }
```

Also it's not referenced in this function. Can we remove the RerouteController.RoutesCallback since we add/remove subscribers? 
```
    override fun reroute(routesCallback: RerouteController.RoutesCallback) {
        state = RerouteState.FetchingRoute
        logger.d(
            Tag(TAG),
            Message("Reroute has been started")
        )
        routeOptionsProvider.update(
            directionsSession.getRouteOptions(),
            tripSession.getRouteProgress(),
            tripSession.getEnhancedLocation()
        )
            ?.let { routeOptions ->
                tryReroute(routeOptions)
            }
            ?: kotlin.run {
                state = RerouteState.Failed("Cannot combine route options")
                state = RerouteState.Idle
            }
    }

Reply:
it's ignored in default implementation only, because `DirectionSession` make routes updates automatically, see https://github.com/mapbox/mapbox-navigation-android/pull/3184/files#diff-e3ca7b557efaeed833ca6f8f76dc2d03R594-R597

what do you mean 🤔 
> since we add/remove subscribers?

Reply:
When I'm searching for routesCallback, it's not referenced. It's not used. 

![Screen Shot 2020-06-30 at 2 04 12 PM](https://user-images.githubusercontent.com/3021882/86176948-a4476280-bada-11ea-9137-6a91fccce326.png)

Reply:
yeah, that's right! it's not used by default implementation, but useful for customer one ☝️

Reply:
@kmadsen what I meant https://github.com/mapbox/mapbox-navigation-android/pull/3184/files#diff-390a6a633e94acf005d69906be84e0ebR377